wireapp/wire-ios-mono
# feat: establish mls one to one group WPB-2182

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-2182" title="WPB-2182" target="_blank"><img alt="Story" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10815?size=medium" />WPB-2182</a>  [iOS] [Clients] Select protocol for 1-1 conversation
  </td></table>
  <br />
 




















































----

## What's in this PR?

This PR establishes mls one to one conversations when a connection is accepted.

Given that A sent a connection to request to B, when B accepts:
- the common supported protocols between the two users is determined.
- If there are no common protocols, the conversation is marked read only.
- If only proteus is supported, then nothing more is done because the proteus one to one conversation already exists.
- If mls is supported, then B will establish the mls group via core crypto and add A to the group.
- A will receive an mls welcome message and process it as normal, then resolve to use the mls conversation for the one on one.
- The old proteus one to one conversation is no longer used for the connection and it is hidden from the list (since it's now considered invalid)

Changes:
- `OneOnOneResolver` resolves which one on one conversation to use for a specific user. Internally it will find the common protocol, update the local state. This is currently called after accepting the connection and receiving a welcome message. 
- `OneOnOneProtocolSelector` resolves which message protocol to use with a specific user.
- `OneOnOneMigrator` migrates a one on one conversation to be used for mls. It will establish the mls group if needed and add the other user to the group, as well as switch the mls conversation to be used for the connection.

## Tests
 
- [x] Unit tests for `OneOnOneResolver`
- [x] Unit tests for `OneOnOneProtocolSelector`
- [x] Unit tests for `OneOnOneMigrator`
- [x] Unit test asserting one on one is resolved after accepting connection
- [x] Unit test asserting one on one is resolved after processing a welcome message
- [x] Unit test asserting invalid one on ones aren't shown in list

## How to test
- User A logs in with a single client to an mls enabled backend and team.
- User B logs in with a single client to the same backend but from a different mls enabled team.
- A sends a connection request to B.
- B accepts the request.
- Assert that both sides see a single mls one to one conversation and can send messages.
